File: packages/core/src/budget.ts
import { openDb } from "./db.js";
import { loadEnv } from "./env.js";

function utcDay(): string {
  return new Date().toISOString().slice(0, 10);
}

export function recordUsage(u: { model: string; stage: string; tokensIn: number; tokensOut: number }): void {
  const db = openDb();
  db.prepare(
    `INSERT INTO usage (day, model, stage, tokens_in, tokens_out, calls)
     VALUES (?, ?, ?, ?, ?, 1)
     ON CONFLICT(day, model, stage) DO UPDATE SET
       tokens_in = tokens_in + excluded.tokens_in,
       tokens_out = tokens_out + excluded.tokens_out,
       calls = calls + 1`,
  ).run(utcDay(), u.model, u.stage, u.tokensIn, u.tokensOut);
}

export function tokensToday(): number {
  const row = openDb()
    .prepare("SELECT COALESCE(SUM(tokens_in + tokens_out), 0) AS total FROM usage WHERE day = ?")
    .get(utcDay()) as { total: number };
  return row.total;
}

export function isOverBudget(): boolean {
  return tokensToday() >= loadEnv().DAILY_TOKEN_BUDGET;
}

export type CostRow = { model: string; stage: string; tokens_in: number; tokens_out: number; calls: number };
export function todayBreakdown(): CostRow[] {
  return openDb()
    .prepare(
      `SELECT model, stage, tokens_in, tokens_out, calls FROM usage WHERE day = ? ORDER BY tokens_in DESC`,
    )
    .all(utcDay()) as CostRow[];
}

File: core/budget.py
from datetime import datetime, timezone


def _utc_day() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d")


def record_usage(*, model: str, stage: str, tokens_in: int, tokens_out: int) -> None:
    from .db import open_db
    db = open_db()
    db.execute(
        """INSERT INTO usage (day, model, stage, tokens_in, tokens_out, calls)
           VALUES (?, ?, ?, ?, ?, 1)
           ON CONFLICT(day, model, stage) DO UPDATE SET
             tokens_in = tokens_in + excluded.tokens_in,
             tokens_out = tokens_out + excluded.tokens_out,
             calls = calls + 1""",
        (_utc_day(), model, stage, tokens_in, tokens_out),
    )
    db.commit()


def tokens_today() -> int:
    from .db import open_db
    row = open_db().execute(
        "SELECT COALESCE(SUM(tokens_in + tokens_out), 0) AS total FROM usage WHERE day = ?",
        (_utc_day(),),
    ).fetchone()
    return row["total"] if row else 0


def is_over_budget() -> bool:
    from .env import load_env
    return tokens_today() >= load_env().DAILY_TOKEN_BUDGET


def today_breakdown() -> list[dict]:
    from .db import open_db
    rows = open_db().execute(
        "SELECT model, stage, tokens_in, tokens_out, calls FROM usage WHERE day = ? ORDER BY tokens_in DESC",
        (_utc_day(),),
    ).fetchall()
    return [dict(r) for r in rows]
